src/stream.rs

```diff
@@ -0,0 +1,584 @@
+    result: CResult,
+    stream: *mut WriteStream,
+    error_message: *mut c_char
+}
+
+unsafe impl Send for PutStreamResponse {}
+
+// RAII Guard for a PutStreamResponse that ensures the awaiting Julia task will be notified
+// even if this is dropped on a panic.
+pub struct PutStreamResponseGuard {
+    response: &'static mut PutStreamResponse,
+    handle: *const c_void
+}
+
+impl NotifyGuard for PutStreamResponseGuard {
+    fn is_uninitialized(&self) -> bool {
+        self.response.result == CResult::Uninitialized
+    }
+    fn condition_handle(&self) -> *const c_void {
+        self.handle
+    }
+    fn set_error(&mut self, error: impl std::fmt::Display) {
+        self.response.result = CResult::Error;
+        self.response.stream = std::ptr::null_mut();
+        let c_string = CString::new(format!("{}", error)).expect("should not have nulls");
+        self.response.error_message = c_string.into_raw();
+    }
+}
+
+impl PutStreamResponseGuard {
+    unsafe fn new(response_ptr: *mut PutStreamResponse, handle: *const c_void) -> PutStreamResponseGuard {
+        let response = unsafe { &mut (*response_ptr) };
+        response.result = CResult::Uninitialized;
+
+        PutStreamResponseGuard { response, handle }
+    }
+    pub(crate) fn success(self, stream: Box<WriteStream>) {
+        self.response.result = CResult::Ok;
+        self.response.stream = Box::into_raw(stream);
+        self.response.error_message = std::ptr::null_mut();
+    }
+}
+
+impl Drop for PutStreamResponseGuard {
+    fn drop(&mut self) {
+        self.notify_on_drop()
+    }
+}
+
+unsafe impl Send for PutStreamResponseGuard {}
+
+#[no_mangle]
+pub extern "C" fn put_stream(
+    path: *const c_char,
+    compress: *const c_char,
+    config: *const Config,
+    response: *mut PutStreamResponse,
+    handle: *const c_void
+) -> CResult {
+    let response = unsafe { PutStreamResponseGuard::new(response, handle) };
+    let path = unsafe { std::ffi::CStr::from_ptr(path) };
+    let path: Path = path.to_str().expect("invalid utf8").try_into().unwrap();
+    let compress = match Compression::try_from(compress) {
+        Ok(c) => c,
+        Err(e) => {
+            // TODO handle early errors with response.set_error
+            std::mem::forget(response);
+            tracing::error!("{}", e);
+            return CResult::Error;
+        }
+    };
+    let config = unsafe { & (*config) };
+
+    match SQ.get() {
+        Some(sq) => {
+            match sq.try_send(Request::PutStream(path, compress, config, response)) {
+                Ok(_) => CResult::Ok,
+                Err(async_channel::TrySendError::Full(_)) => {
+                    CResult::Backoff
+                }
+                Err(async_channel::TrySendError::Closed(_)) => {
+                    CResult::Error
+                }
+            }
+        }
+        None => {
+            return CResult::Error;
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn write_to_stream(
```

Comment:
I think we need some more documentation for these more complicated features, either as comments or in the README.
Does this handle chunking the puts into the size needed by each cloud provider?
Do you need to flush, or will shutdown do it for you?
How is concurrency handled (seems unbounded and needs to be controlled by the user)?

Reply:
I will document some of the more complex functions. But answering some of your questions:
- It splits the bytes written into 10 MB chunks which (according to their code) is allowed by the supported cloud providers.
- Only shutdown is required (it does flushing and some final completion request), flush is only exposed to give the user control over any internal buffering.
- It uploads up to 8 parts in parallel and the user has no control over this (it is hard-coded), also any interaction with the stream from the Julia side is not thread-safe, so no concurrency from a Julia perspective.